### PR TITLE
REST API: Add locations to the menu response

### LINKED ISSUE #31648: REST API: Menu response doesn't return locations
## Description
The menu item response doesn't include assigned menu locations. The schema defines the `locations` property, but it's not included in the response.

## Step-by-step reproduction instructions
I'm using Paw and Application Password for making REST API requests, but any similar tool should work for tests.

1. Request API menu item. `GET /wp-json/__experimental/menus/:id`
2. See that `locations` property is missing.

Check schema using `OPTIONS /wp-json/__experimental/menus`.
 
## Expected behaviour
The response has a `locations` property with assigned menu locations.

## Actual behaviour
The `locations` property is missing.

## WordPress information
- WordPress version: 5.7.1
- Gutenberg version: 10.6.0 RC1
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? Yes

### PR DESCRIPTION
**Pull request by Mamaduka**
## Description
Fixes #31648.

## How has this been tested?
1. Send a request to get a menu item. GET `/wp-json/__experimental/menus/:id`
2. See that response has a `locations` property with assigned menu locations.

## Types of changes
 Bugfix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: lib/class-wp-rest-menus-controller.php [hunk 1]
```diff
@@ -189,6 +189,10 @@ public function prepare_item_for_response( $term, $request ) {
 		$fields = $this->get_fields_for_response( $request );
 		$data   = $response->get_data();
 
+		if ( in_array( 'locations', $fields, true ) ) {
+			$data['locations'] = $this->get_menu_locations( $nav_menu->term_id );
+		}
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menus-controller.php:194]: Why is this needed? You can get menu locations from embedding request?

`?_embed=menu-location` at the end of the request.

Comment (spacedmonkey) [reply lib/class-wp-rest-menus-controller.php:194]: Also, you can just add locations like this. You would need to schema.

File: lib/class-wp-rest-menus-controller.php [hunk 2]
```diff
@@ -531,6 +533,26 @@ function handle_auto_add( $menu_id, $request ) {
 		return $update;
 	}
 
+	/**
+	 * Returns names of the locations assigned to the menu.
+	 *
+	 * @param int $menu_id The menu id.
```
Comment (spacedmonkey) [inline lib/class-wp-rest-menus-controller.php:541]: Add a since and a unit test.

File: lib/class-wp-rest-menus-controller.php (hunk 1 above)
Comment (spacedmonkey) [reply lib/class-wp-rest-menus-controller.php:194]: Screenshot 2021-05-10 at 12 17 49

Comment (Mamaduka) [reply lib/class-wp-rest-menus-controller.php:194]: Primarily because of consistency when using with entities on the JS side. You can set locations using `editEntityRecord`, but the `getEntityRecord` doesn't return them.

The `locations` is also defined in the schema but never returned.

> ?_embed=menu-location at the end of the request.

I just tested this, and it doesn't embed menu locations for some reason.

File: lib/class-wp-rest-menus-controller.php (hunk 2 above)
Comment (Mamaduka) [reply lib/class-wp-rest-menus-controller.php:541]: Won't be `since` tags updated if/when this endpoint gets merged into WP core?

I will work on unit tests.

File: lib/class-wp-rest-menus-controller.php (hunk 1 above)
Comment (Mamaduka) [reply lib/class-wp-rest-menus-controller.php:194]: In a way, this is similar to how a post response has an array of category/tag IDs.

File: lib/class-wp-rest-menus-controller.php (hunk 2 above)
Comment (Mamaduka) [reply lib/class-wp-rest-menus-controller.php:541]: Added unit test.

Comment (Mamaduka) [reply lib/class-wp-rest-menus-controller.php:541]: Added since.

**Comment by TimothyBJacobs** [comment]
IIRC we intentionally omitted the locations from the menu object. The locations are a property of the theme, and should be managed through the `menu-locations` endpoint. Is the menu locations endpoint not working properly?

**Comment by Mamaduka** [comment]
Currently, you can only get locations using the `menu-locations` endpoint. To assign a menu to a location, you need to use the `menus` endpoint.

The primary reason I created this PR was inconsistency with `getEntityRecord` and `editEntityRecord`. When latter is used to set locations, but the former wasn't returning them.

It's also confusing to have an item in the schema but not in the response IMO.

**Comment by TimothyBJacobs** [comment]
Huh, I could have sworn it was the opposite, that updating was done thru the menu locations endpoint. If we're doing updating in the menus endpoint, then yes we definitely need to surface that value in the response.

**Comment by Mamaduka** [comment]
Do you think we should fully move menu location management to the `menu-locations` endpoint or leave it as it is?

**Comment by TimothyBJacobs** [comment]
Yes, my preference would be to put it in the locations endpoint, but I don't think it's a deal breaker.

Actually, thinking about it more. I think it'd probably make the most sense to have the menu locations be part of the theme's endpoint itself and use that for both retrieving and editing.

**Comment by Mamaduka** [comment]
Looking at the `widgets` and `sidebars` endpoints, they both handle assigning widgets to a sidebar. The sidebars are also a property of the theme.

Maybe we should use the same logic for `menus` and `menu-locations.`

**Comment by Mamaduka** [comment]
Hi, @spacedmonkey, @TimothyBJacobs

Is it okay to move forward with this solution?

Thanks

**Comment by TimothyBJacobs** [comment]
Yeah, the sidebars endpoint is definitely something to look at. I think a difference would be the complexity of data, the sidebars have a lot of complex things to manage, while the menu locations are fairly straightforward and are basically just a wrapper around the theme mod api. But let's keep things simple and roll with this PR for now.

**Review by TimothyBJacobs — APPROVED** [review]
